utils/dwarf.c

```diff
@@ -996,6 +998,100 @@ static int get_dwarfspecs_cb(Dwarf_Die *die, void *data)
 	return DWARF_CB_OK;
 }
 
+struct dir_entry {
+	struct rb_node	node;
+	char		*name;
+	int		nr_locs;
+	int		nr_used;
+};
+
+static int add_dir_entry(struct rb_root *root, char *name, int nr_locs)
+{
+	struct dir_entry *entry, *iter;
+	struct rb_node *parent = NULL;
+	struct rb_node **p = &root->rb_node;
+	int cmp;
+
+	pr_dbg3("add dir entry: %s (%d)\n", name, nr_locs);
+
+	while (*p) {
+		parent = *p;
+		iter = rb_entry(parent, struct dir_entry, node);
+
+		cmp = strcmp(iter->name, name);
+		if (cmp == 0) {
+			iter->nr_used++;
+			iter->nr_locs += nr_locs;
+			return 0;
+		}
+
+		if (cmp > 0)
+			p = &parent->rb_left;
+		else
+			p = &parent->rb_right;
+	}
+
+	entry = xmalloc(sizeof(*entry));
+	entry->name = name;
+	entry->nr_locs = nr_locs;
+	entry->nr_used = 1;
+
+	rb_link_node(&entry->node, parent, p);
+	rb_insert_color(&entry->node, root);
+
+	return 0;
+}
+
+static void free_dir_entry(struct rb_root *root)
+{
+	struct dir_entry *entry;
+	struct rb_node *node;
+
+	while (!RB_EMPTY_ROOT(root)) {
+		node = rb_first(root);
+		entry = rb_entry(node, typeof(*entry), node);
+
+		rb_erase(node, root);
+		free(entry);
+	}
+}
+
+char* get_base_dir(struct rb_root *dirs)
```

Comment:
I think the names are too generic.  Also this can be a static function..

How about below?
* add_comp_dir()
* free_comp_dir()
* get_base_comp_dir()